src/middlewared/middlewared/plugins/device_/vrrp_events_linux.py

```diff
@@ -0,0 +1,25 @@
+import os
+
+from middlewared.utils import start_daemon_thread
+
+VRRP_FIFO_FILE = '/var/run/vrrpd.fifo'
+
+
+def vrrp_fifo_listen(middleware):
+
+    # create the fifo, ignoring if it already exists
+    try:
+        os.mkfifo(VRRP_FIFO_FILE)
+    except FileExistsError:
+        pass
+
+    while True:
+        with open(VRRP_FIFO_FILE) as f:
+            middleware.logger.info('vrrp fifo connection established')
+            # all vrrp messages are terminated with a newline
+            for line in f:
+                middleware.call_hook_sync('vrrp.fifo', data=line)
```

Comment:
Wonder if we should be extra safe here and wrap it all around a try/except so it keeps retrying after X seconds if for any reason we get an exception

Reply:
I've tried hard to ensure that the FIFO is created in this method before the keepalived service is started. If the keepalived service starts before middlewared, and it has created the FIFO then the FIFO file gets removed which would cause an exception. If middlewared creates the FIFO, then the file persists even if middlewared is restarted which is what we want. However, I can add an exception for safe measure.